File: agentarea-platform/libs/common/agentarea_common/events/contract.py
"""Canonical event contract: dotted taxonomy and supersede-by-id parts.

Pure, side-effect-free. No broker, redis, or db imports. Both the backend
read-path decoder and the frontend reducer mirror this contract.

The source emits the canonical dotted names directly (see EventTypes in
agentarea_execution.workflows.constants); there is no legacy vocabulary and no
alias-on-read bridge.

Core idea — supersede-by-id: every non-lifecycle event maps to a Part with a
stable ``part_id``; a later event with the same ``part_id`` replaces that part
in place. Lifecycle/terminal ``task.*`` events are append-only and derive no
Part.
"""

from __future__ import annotations

from collections.abc import Iterable
from dataclasses import dataclass
from typing import Any

# Canonical (dotted) types — the one vocabulary the source emits and the wire
# carries.
LLM_STARTED = "llm.call.started"
LLM_COMPLETED = "llm.call.completed"
LLM_FAILED = "llm.call.failed"
LLM_CHUNK = "llm.call.chunk"
TOOL_CALL = "tool.call"
TOOL_RESULT = "tool.result"
INPUT_REQUEST = "input.request"
INPUT_RESPONSE = "input.response"
APPROVAL_REQUEST = "approval.request"
APPROVAL_RESPONSE = "approval.response"
ARTIFACT_CREATED = "artifact.created"
ARTIFACT_UPDATED = "artifact.updated"
A2UI_CREATE = "a2ui.create"
A2UI_UPDATE_COMPONENTS = "a2ui.update.components"
A2UI_UPDATE_DATA = "a2ui.update.data"
A2UI_DELETE = "a2ui.delete"
TASK_STARTED = "task.started"
TASK_COMPLETED = "task.completed"
TASK_FAILED = "task.failed"
TASK_CANCELLED = "task.cancelled"

# Canonical terminal types: append-only lifecycle events that end a task feed.
_TERMINAL_TYPES: frozenset[str] = frozenset({TASK_COMPLETED, TASK_FAILED, TASK_CANCELLED})


def canonical_type(event_type: str) -> str:
    """Return the canonical dotted event type (identity for canonical inputs).

    The source already emits canonical names, so this only strips a leading
    ``workflow.`` prefix defensively. Kept as the call-site seam so read-path
    comparison sites stay uniform.
    """
    return event_type[len("workflow.") :] if event_type.startswith("workflow.") else event_type


# Canonical type -> part kind. Absence means "no part" (lifecycle/terminal).
_KIND_BY_TYPE: dict[str, str] = {
    LLM_STARTED: "llm",
    LLM_COMPLETED: "llm",
    LLM_FAILED: "llm",
    LLM_CHUNK: "llm",
    TOOL_CALL: "tool",
    TOOL_RESULT: "tool",
    INPUT_REQUEST: "form",
    INPUT_RESPONSE: "form",
    APPROVAL_REQUEST: "form",
    APPROVAL_RESPONSE: "form",
    ARTIFACT_CREATED: "artifact",
    ARTIFACT_UPDATED: "artifact",
    A2UI_CREATE: "a2ui",
    A2UI_UPDATE_COMPONENTS: "a2ui",
    A2UI_UPDATE_DATA: "a2ui",
    A2UI_DELETE: "a2ui",
}


@dataclass(frozen=True)
class Part:
    part_id: str
    kind: str
    event_type: str
    data: dict[str, Any]


def _part_id_for(kind: str, canonical: str, data: dict[str, Any]) -> str | None:
    if kind == "tool":
        return _as_str(data.get("tool_call_id"))
    if kind == "llm":
        execution_id = data.get("execution_id")
        iteration = data.get("iteration")
        if execution_id is None or iteration is None:
            return None
        return f"{execution_id}:{iteration}"
    if kind == "form":
        if canonical in (APPROVAL_REQUEST, APPROVAL_RESPONSE):
            return _as_str(data.get("escalation_id") or data.get("request_id"))
        return _as_str(data.get("input_request_id") or data.get("request_id"))
    if kind == "a2ui":
        return _as_str(data.get("surface_id"))
    if kind == "artifact":
        return _as_str(data.get("artifact_id"))
    return None


def _as_str(value: Any) -> str | None:
    if value is None:
        return None
    text = str(value)
    return text or None


def derive_part(event_type: str, data: dict[str, Any]) -> Part | None:
    """Derive the superseding Part for an event, or None for lifecycle events."""
    canonical = canonical_type(event_type)
    kind = _KIND_BY_TYPE.get(canonical)
    if kind is None:
        return None
    part_id = _part_id_for(kind, canonical, data)
    if part_id is None:
        return None
    return Part(part_id=part_id, kind=kind, event_type=canonical, data=data)


def ensure_terminal_message(event_type: str, data: dict[str, Any]) -> dict[str, Any]:
    """Return terminal event data with a user-facing ``message`` (additive).

    A client attaching AFTER a task finishes renders the final state from the
    catch-up snapshot alone, so terminal events must carry a human-readable
    ``message`` (and ``reason`` for failed/cancelled). Non-terminal events pass
    through unchanged, and an existing ``message`` is never overwritten.

    Returns the same dict instance mutated in place; the caller may pass a copy
    if the source must stay untouched.
    """
    canonical = canonical_type(event_type)
    if canonical not in _TERMINAL_TYPES:
        return data

    if canonical == TASK_COMPLETED:
        if not data.get("message"):
            final = _as_str(data.get("final_response") or data.get("result"))
            data["message"] = final or "Task completed."
        return data

    # Failed / cancelled: carry a reason too.
    reason = _as_str(
        data.get("reason")
        or data.get("error")
        or data.get("blocked_reason")
        or data.get("error_type")
    )
    if canonical == TASK_FAILED:
        reason = reason or "Task failed."
        data.setdefault("reason", reason)
        if not data.get("message"):
            data["message"] = reason
    else:  # TASK_CANCELLED
        reason = reason or "Task cancelled."
        data.setdefault("reason", reason)
        if not data.get("message"):
            data["message"] = reason
    return data


def reduce_parts(events: Iterable[tuple[str, dict[str, Any]]]) -> list[Part]:
    """Fold an event stream into ordered parts via supersede-by-id.

    Non-part events are skipped. A later part with an existing ``part_id``
    replaces the earlier one at its original position.
    """
    order: list[str] = []
    by_id: dict[str, Part] = {}
    for event_type, data in events:
        part = derive_part(event_type, data)
        if part is None:
            continue
        if part.part_id not in by_id:
            order.append(part.part_id)
        by_id[part.part_id] = part
    return [by_id[pid] for pid in order]

File: agentarea-webapp/src/lib/events/contract.ts
/**
 * Canonical event contract: dotted taxonomy and supersede-by-id parts.
 *
 * Pure, side-effect-free mirror of the backend contract
 * (agentarea_common/events/contract.py). The source emits the canonical dotted
 * names directly — there is no second vocabulary and no alias-on-read bridge.
 *
 * The whole point is supersede-by-id: every non-lifecycle event maps to a Part
 * with a stable partId; a later event with the same partId replaces that part
 * in place. Lifecycle/terminal task.* events derive no Part.
 */

export const LLM_STARTED = "llm.call.started";
export const LLM_COMPLETED = "llm.call.completed";
export const LLM_FAILED = "llm.call.failed";
export const LLM_CHUNK = "llm.call.chunk";
export const TOOL_CALL = "tool.call";
export const TOOL_RESULT = "tool.result";
export const INPUT_REQUEST = "input.request";
export const INPUT_RESPONSE = "input.response";
export const APPROVAL_REQUEST = "approval.request";
export const APPROVAL_RESPONSE = "approval.response";
export const ARTIFACT_CREATED = "artifact.created";
export const ARTIFACT_UPDATED = "artifact.updated";
export const A2UI_CREATE = "a2ui.create";
export const A2UI_UPDATE_COMPONENTS = "a2ui.update.components";
export const A2UI_UPDATE_DATA = "a2ui.update.data";
export const A2UI_DELETE = "a2ui.delete";
export const TASK_COMPLETED = "task.completed";
export const TASK_FAILED = "task.failed";
export const TASK_CANCELLED = "task.cancelled";
export const TASK_AWAITING_CONTINUATION = "task.awaiting_continuation";
export const TASK_CONTINUED = "task.continued";

export const TERMINAL_TYPES: ReadonlySet<string> = new Set([
  TASK_COMPLETED,
  TASK_FAILED,
  TASK_CANCELLED,
]);

/**
 * Return the canonical dotted event type (identity for canonical inputs). The
 * source already emits canonical names, so this only strips a leading
 * `workflow.` prefix defensively.
 */
export function canonicalType(eventType: string): string {
  return eventType.startsWith("workflow.")
    ? eventType.slice("workflow.".length)
    : eventType;
}

export type PartKind = "llm" | "tool" | "form" | "artifact" | "a2ui";

const KIND_BY_TYPE: Record<string, PartKind> = {
  [LLM_STARTED]: "llm",
  [LLM_COMPLETED]: "llm",
  [LLM_FAILED]: "llm",
  [LLM_CHUNK]: "llm",
  [TOOL_CALL]: "tool",
  [TOOL_RESULT]: "tool",
  [INPUT_REQUEST]: "form",
  [INPUT_RESPONSE]: "form",
  [APPROVAL_REQUEST]: "form",
  [APPROVAL_RESPONSE]: "form",
  [ARTIFACT_CREATED]: "artifact",
  [ARTIFACT_UPDATED]: "artifact",
  [A2UI_CREATE]: "a2ui",
  [A2UI_UPDATE_COMPONENTS]: "a2ui",
  [A2UI_UPDATE_DATA]: "a2ui",
  [A2UI_DELETE]: "a2ui",
};

export type EventData = Record<string, unknown>;

export interface Part {
  partId: string;
  kind: PartKind;
  eventType: string;
  data: EventData;
}

function asStr(value: unknown): string | null {
  if (value === null || value === undefined) return null;
  const text = String(value);
  return text || null;
}

function partIdFor(
  kind: PartKind,
  canonical: string,
  data: EventData
): string | null {
  if (kind === "tool") {
    return asStr(data.tool_call_id);
  }
  if (kind === "llm") {
    const executionId = data.execution_id;
    const iteration = data.iteration;
    if (
      executionId === null ||
      executionId === undefined ||
      iteration === null ||
      iteration === undefined
    ) {
      return null;
    }
    return `${executionId}:${iteration}`;
  }
  if (kind === "form") {
    if (canonical === APPROVAL_REQUEST || canonical === APPROVAL_RESPONSE) {
      return asStr(data.escalation_id ?? data.request_id);
    }
    return asStr(data.input_request_id ?? data.request_id);
  }
  if (kind === "a2ui") {
    return asStr(data.surface_id);
  }
  return asStr(data.artifact_id);
}

/** Derive the superseding Part for an event, or null for lifecycle events. */
export function derivePart(eventType: string, data: EventData): Part | null {
  const canonical = canonicalType(eventType);
  const kind = KIND_BY_TYPE[canonical];
  if (!kind) return null;
  const partId = partIdFor(kind, canonical, data);
  if (partId === null) return null;
  return { partId, kind, eventType: canonical, data };
}

export interface EventInput {
  eventType: string;
  data: EventData;
}

/**
 * Fold an event stream into ordered parts via supersede-by-id. Non-part events
 * are skipped. A later part with an existing partId replaces the earlier one at
 * its original position.
 */
export function reduceParts(events: Iterable<EventInput>): Part[] {
  const order: string[] = [];
  const byId = new Map<string, Part>();
  for (const { eventType, data } of events) {
    const part = derivePart(eventType, data);
    if (part === null) continue;
    if (!byId.has(part.partId)) order.push(part.partId);
    byId.set(part.partId, part);
  }
  return order.map((pid) => byId.get(pid) as Part);
}
